File: sorting/python/merge_sort.py
# -*- coding: utf-8 -*

'''
    并归排序
    通过递归的形式实现
'''
import sort_helper
import sys
sys.setrecursionlimit(10000000)
def __merage_sort(arr,l,r):
    if l >= r:
        return

    mid = l + int((r - l) / 2)
    __merage_sort(arr,l,mid)
    __merage_sort(arr,mid+1,r)
    __merage(arr,l,mid,r)

# 进行归并排序
def __merage(arr,l,mid,r):
    aux = arr[l:r+1]
    i = l
    j = mid + 1
    for k in range(l,r+1):
        if i > mid:
            arr[k] = aux[j - l]
            j +=1
        elif j > r:
            arr[k] = aux[i - l]
            i +=1
        elif aux[i-l] < aux[j - l]:
            arr[k] = aux[i-l]
            i +=1
        else:
            arr[k] = aux[j - l]
            j +=1

@sort_helper.test_sort
def sort(arr):
    __merage_sort(arr,0,len(arr)-1)

if __name__ == '__main__':
    a = sort_helper.generate_randoma_array(100000,1,99999)
    sort(a)

File: sorting/golang/merge_sort.go
package main

import (
	"./help"
)

func merge_sort(arr []int, l int, r int) {

	if l >= r {
		return
	}
	mid := l + (r-l)/2
	merge_sort(arr, l, mid)
	merge_sort(arr, mid+1, r)
	merge(arr, l, mid, r)
}

func merge(arr []int, l int, mid int, r int) {
	// copy 临时array
	aux := make([]int, r-l + 1)
	copy(aux,arr[l : r+1])

	i, j := l, mid+1
	for k := l; k < r+1; k++ {
		if i > mid {
			arr[k] = aux[j-l]
			j++
		} else if j > r {
			arr[k] = aux[i-l]
			i++
		} else if aux[i-l] < aux[j-l] {
			arr[k] = aux[i-l]
			i++
		} else {
			arr[k] = aux[j-l]
			j++
		}
	}

}

func sort(arr []int, n int) []int {
	merge_sort(arr, 0, n-1)
	return arr
}

func main() {
	arr := help.GenRanArray(1000000, 1, 999999)
	//help.ArrayPrint(arr)
	help.TestSort(sort)(arr, len(arr))
	//help.ArrayPrint(arr)

}
